# Applied Data Science Capstone Project: Segmenting and Clustering Neighborhoods in Toronto-Problem 2

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.


In [43]:
# Import necessary libraries

import requests
import lxml.html as lh
import pandas as pd

In [44]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle, page, to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [45]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This means that there are 3 columns per row

In [47]:
# Parse the first row as our header
tr_elements = doc.xpath('//tr')

#Create empty list
col=[]
i=0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postcode"
2:"Borough"
3:"Neighborhood
"


**Creating Pandas DataFrame** <br>
Each header is appended to a tuple along with an empty list.

In [48]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [49]:
# Check the length of each column. Ideally, they should all be the same
[len(C) for (title,C) in col]

[287, 287, 287]

This shows that each of the 3 columns has exactly 287 rows

**Creating the pandas data frame**

In [50]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

In [51]:
# Access the top 5 rows of the data frame 
df.head()

Postcode           Borough      Neighborhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

**Rearranging and renaming the columns**

In [54]:
df.columns = ['Borough', 'Neighbourhood','Postcode']

cols = df.columns.tolist()
cols

cols = cols[-1:] + cols[:-1]

df = df[cols]

df.head()

Postcode           Borough       Neighbourhood
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

**Cleaning the messy string in the Borough column**

In [55]:
df = df.replace('\n',' ', regex=True)
df.head()

Postcode           Borough      Neighbourhood
0      M1A      Not assigned      Not assigned 
1      M2A      Not assigned      Not assigned 
2      M3A        North York         Parkwoods 
3      M4A        North York  Victoria Village 
4      M5A  Downtown Toronto      Harbourfront

**Dropping all cells with a borough that is Not assigned**

In [56]:
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop=True)

df.head(10)

Postcode           Borough      Neighbourhood
0      M3A        North York         Parkwoods 
1      M4A        North York  Victoria Village 
2      M5A  Downtown Toronto      Harbourfront 
3      M6A        North York  Lawrence Heights 
4      M6A        North York    Lawrence Manor 
5      M7A      Queen's Park      Not assigned 
6      M9A  Downtown Toronto      Queen's Park 
7      M1B       Scarborough             Rouge 
8      M1B       Scarborough           Malvern 
9      M3B        North York   Don Mills North

**Combining Neighbourhoods based on similar Postcode and Borough**

In [57]:
df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head(10)

Postcode      Borough                                     Neighbourhood
0      M1B  Scarborough                                   Rouge ,Malvern 
1      M1C  Scarborough           Highland Creek ,Rouge Hill ,Port Union 
2      M1E  Scarborough                Guildwood ,Morningside ,West Hill 
3      M1G  Scarborough                                           Woburn 
4      M1H  Scarborough                                        Cedarbrae 
5      M1J  Scarborough                              Scarborough Village 
6      M1K  Scarborough      East Birchmount Park ,Ionview ,Kennedy Park 
7      M1L  Scarborough                  Clairlea ,Golden Mile ,Oakridge 
8      M1M  Scarborough  Cliffcrest ,Cliffside ,Scarborough Village West 
9      M1N  Scarborough                      Birch Cliff ,Cliffside West

In [58]:
#Removing any space in the start of the string
df['Neighbourhood'] = df['Neighbourhood'].str.strip()

In [59]:
#Assigning Borough values to the Neighbourhood where vlaue is "Not assigned"
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [60]:
# Check if the Neighbourhood for Queen's Park changed 
df[df['Borough'] == 'Queen\'s Park']

Postcode       Borough Neighbourhood
85      M7A  Queen's Park  Queen's Park

In [61]:
# Check the shape of the data frame
df.shape

(103, 3)

In [62]:
#Save this file to a csv
df.to_csv(r'df_can.csv')

In [63]:
#Extracting the Latitude and Longitude of Canada Ontario
"""import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')"""

"import requests # library to handle requests\nimport pandas as pd # library for data analsysis\nimport numpy as np # library to handle data in a vectorized manner\nimport random # library for random number generation\n\n!conda install -c conda-forge geopy --yes \nfrom geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values\n\n# libraries for displaying images\nfrom IPython.display import Image \nfrom IPython.core.display import HTML \n    \n# tranforming json file into a pandas dataframe library\nfrom pandas.io.json import json_normalize\n\n!conda install -c conda-forge folium=0.5.0 --yes\nimport folium # plotting library\n\nprint('Folium installed')\nprint('Libraries imported.')"

In [64]:
"""import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]"""

"import geocoder # import geocoder\n\n# initialize your variable to None\nlat_lng_coords = None\n\n# loop until you get the coordinates\nwhile(lat_lng_coords is None):\n  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))\n  lat_lng_coords = g.latlng\n\nlatitude = lat_lng_coords[0]\nlongitude = lat_lng_coords[1]"

### Couldn't figure out the error, So I used the csv given in the link

In [72]:
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner

link = "http://cocl.us/Geospatial_data"
df1 = pd.read_csv(link)

df1.head()

HTTPError: HTTP Error 503: Service Unavailable

In [73]:
df1.shape

(103, 3)

Both the data frames have 103 rows and 3 columns

**Changing the column name Postal code to Postcode to merge the two data frames together**

In [74]:
df1.columns = ['Postcode','Latitude','Longitude']

cols = df1.columns.tolist()
cols

['Postcode', 'Latitude', 'Longitude']

In [75]:
#Read in the CSV file saved in the previous assignment
df = pd.read_csv(r'df_can.csv')
df.head()

Unnamed: 0 Postcode      Borough                           Neighbourhood
0           0      M1B  Scarborough                          Rouge ,Malvern
1           1      M1C  Scarborough  Highland Creek ,Rouge Hill ,Port Union
2           2      M1E  Scarborough       Guildwood ,Morningside ,West Hill
3           3      M1G  Scarborough                                  Woburn
4           4      M1H  Scarborough                               Cedarbrae

**Merging the two data frames together based on their Postcode**